src/java.base/share/classes/java/nio/file/CopyMoveHelper.java

```diff
@@ -131,6 +131,10 @@ static void copyToForeignTarget(Path source, Path target,
 
         // delete target if it exists and REPLACE_EXISTING is specified
         if (opts.replaceExisting) {
+            // First ensure source can be read
+            source.getFileSystem().provider().checkAccess(source,
+                                                          AccessMode.READ);
```

Comment:
This looks right for the cross-provider case. Minor nit is that you don't ned the line break here. Also the existing comments started in lower case.

Reply:
Fixed in e226096d15b47bee7cee20ec2994bca93a071433.